### PR TITLE
fix: auto subtitles

### PR DESCRIPTION
**Pull request by Kneesal**
## Summary by CodeRabbit

* **New Features**
  * Introduced a comprehensive language switcher dialog for site, audio, and subtitle language selection, accessible from header and video controls.
  * Added locale-aware fetching and display of video audio and subtitle languages with automatic subtitle enabling based on user preferences.
  * Enhanced video browsing and playback with localized titles, descriptions, and improved internationalization support.
  * Enabled filtering and display of videos and subtitles based on selected or detected language.

* **Improvements**
  * Added suggested language prioritization based on site, browser, and country in language selection components.
  * Improved accessibility and localization of UI elements including dialog labels, buttons, and content type labels.
  * Refined logic for prioritizing discussion questions and selecting video metadata.
  * Enhanced subtitle management with automatic switching, track updates, and synchronization during playback.

* **Bug Fixes**
  * Fixed synchronization issues between language and subtitle states across components.
  * Improved handling of edge cases in language and subtitle selection and URL redirection logic.

* **Tests**
  * Added extensive unit and integration tests covering language selection dialogs, audio and subtitle management, context state, and locale-based middleware.

* **Chores**
  * Centralized subtitle language IDs and locale mappings for consistency.
  * Refactored language-related utilities and hooks for better modularity and side-effect handling.
  * Updated localization files with new keys for language options, video playback, downloads, sharing, and content categorization.

### CODE REVIEW

File: apps/watch/pages/watch/[part1]/[part2]/[part3].tsx [hunk 1]
```diff
@@ -67,7 +75,8 @@ export default function Part3Page({
     subtitleLanguage: getCookie('SUBTITLE_LANGUAGE') ?? '529',
     subtitleOn: (getCookie('SUBTITLES_ON') ?? 'false') === 'true',
     videoId: content.id,
-    videoVariantSlug: content.variant?.slug
+    videoVariantSlug: content.variant?.slug,
+    videoSubtitleLanguages
```
Comment (csiyang) [inline apps/watch/pages/watch/[part1]/[part2]/[part3].tsx:79]: hmm weird this isnt throwing a type error, I used a different type for inital state. Can we type the const as WatchInitialState

File: apps/watch/pages/watch/[part1]/[part2]/[part3].tsx [hunk 2]
```diff
@@ -51,11 +57,13 @@ export const GET_VIDEO_CONTENT_PART_3 = gql`
...
   container: VideoContentFields
   content: VideoContentFields
+  videoSubtitleLanguages: VideoVariantSubtitle[]
 }
 
 export default function Part3Page({
   container,
-  content
+  content,
+  videoSubtitleLanguages
```
Comment (csiyang) [inline apps/watch/pages/watch/[part1]/[part2]/[part3].tsx:66]: We need to do this for [part2] as well

**Review by csiyang — CHANGES_REQUESTED** [review]

**Comment by Kneesal** [comment]
@coderabbitai resolve

File: apps/watch/src/components/LanguageSwitchDialog/SubtitlesSelect/SubtitlesSelect.tsx [hunk 3]
```diff
@@ -134,7 +134,16 @@ export const SubtitlesSelect = memo(function SubtitlesSelect(): ReactElement {
           id="no-subtitles"
           type="checkbox"
           checked={preferredSubtitleOn}
-          onChange={() => updateSubtitlesOn(!preferredSubtitleOn)}
+          onChange={() => {
+            if (autoSubtitle) {
```
Comment (csiyang) [inline apps/watch/src/components/LanguageSwitchDialog/SubtitlesSelect/SubtitlesSelect.tsx:138]: We should have a comment here explaining why we're doing this

File: apps/watch/src/components/LanguageSwitchDialog/SubtitlesSelect/SubtitlesSelect.tsx [hunk 4]
```diff
@@ -55,7 +55,7 @@ export const SubtitlesSelect = memo(function SubtitlesSelect(): ReactElement {
     }
   }, [videoId, videoSubtitleLanguages, getSubtitleLanguages])
 
-  const preferredSubtitleOn = autoSubtitle ?? subtitleOn
+  const preferredSubtitleOn = autoSubtitle === true ? true : subtitleOn
```
Comment (csiyang) [inline apps/watch/src/components/LanguageSwitchDialog/SubtitlesSelect/SubtitlesSelect.tsx:58]: Does the undefined check not work anymore?

**Review by csiyang — CHANGES_REQUESTED** [review]

File: apps/watch/src/components/LanguageSwitchDialog/SubtitlesSelect/SubtitlesSelect.tsx (hunk 4 above)
Comment (Kneesal) [reply apps/watch/src/components/LanguageSwitchDialog/SubtitlesSelect/SubtitlesSelect.tsx:58]: I have reverted this back. It was correct

File: apps/watch/src/components/LanguageSwitchDialog/SubtitlesSelect/SubtitlesSelect.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,728 @@
...
+
+    const checkbox = screen.getByRole('checkbox')
+    await waitFor(() => {
+      expect(checkbox).toBeChecked()
+    })
+    // Test that the checkbox can be clicked (interaction works)
+    await user.click(checkbox)
+    // Since this is an integration test, the action is executed successfully
+    // if no errors are thrown during the click event
+    expect(checkbox).toBeInTheDocument()
```
[the remark this answers is not in the document]
Comment (Kneesal) [reply apps/watch/src/components/LanguageSwitchDialog/SubtitlesSelect/SubtitlesSelect.spec.tsx:451]: the comments need to remain to inform future developers of our intention

File: apps/watch/src/components/LanguageSwitchDialog/SubtitlesSelect/SubtitlesSelect.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,728 @@
...
+        </WatchProvider>
+      </MockedProvider>
+    )
+
+    // Should render main elements
+    expect(screen.getByText('Subtitles')).toBeInTheDocument()
+    expect(screen.getByRole('combobox')).toBeInTheDocument()
+    expect(screen.getByLabelText('Show subtitles')).toBeInTheDocument()
+    expect(screen.getByRole('checkbox')).toBeInTheDocument()
+    expect(document.querySelector('svg')).toBeInTheDocument() // ClosedCaptionOffOutlinedIcon
```
[the remark this answers is not in the document]
Comment (Kneesal) [reply apps/watch/src/components/LanguageSwitchDialog/SubtitlesSelect/SubtitlesSelect.spec.tsx:125]: this is not how we test our icons.

**Review by csiyang — APPROVED** [review]
